### ISSUE TITLE
torch.compile produces inconsistent outputs for MultiHeadAttentionModel with scaled_dot_product_attention

### ISSUE THREAD

**Issue by Blooming-Tree** [pytorch/pytorch#188420]
### 🐛 Describe the bug

When applying torch.compile to a multi-head attention model that uses F.scaled_dot_product_attention, the compiled model produces outputs that significantly deviate from eager execution, exceeding acceptable error thresholds.
code:
```
import torch
import copy
import torch.nn as nn
import torch.nn.functional as F
import os

class MultiHeadAttentionModel(nn.Module):
    def __init__(self, d_model=512, n_heads=8, seq_len=64):
        super().__init__()
        self.d_model = d_model
        self.n_heads = n_heads
        self.head_dim = d_model // n_heads
        self.query_proj = nn.Linear(d_model, d_model)
        self.key_proj = nn.Linear(d_model, d_model)
        self.value_proj = nn.Linear(d_model, d_model)
        self.out_proj = nn.Linear(d_model, d_model)
        self.seq_len = seq_len

    def forward(self, x):
        batch_size = x.size(0)
        query = self.query_proj(x)
        key = self.key_proj(x)
        value = self.value_proj(x)
        query = query.view(batch_size, self.seq_len, self.n_heads, self.head_dim).transpose(1, 2)
        key = key.view(batch_size, self.seq_len, self.n_heads, self.head_dim).transpose(1, 2)
        value = value.view(batch_size, self.seq_len, self.n_heads, self.head_dim).transpose(1, 2)
        attn_output = F.scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.1, is_causal=False)
        attn_output = attn_output.transpose(1, 2).contiguous().view(batch_size, self.seq_len, self.d_model)
        output = self.out_proj(attn_output)
        return output

def get_default_model():
    return MultiHeadAttentionModel(d_model=512, n_heads=8, seq_len=64)

def get_sample_inputs():
    return (torch.randn(4, 64, 512),)

def main():
    torch.use_deterministic_algorithms(True)
    torch.set_float32_matmul_precision('highest')
    os.environ['CUBLAS_WORKSPACE_CONFIG'] = ':4096:8'
    torch._inductor.config.fallback_random = True

    torch.manual_seed(42)
    model = get_default_model()
    model = model.double()
    model.eval()

    inputs = get_sample_inputs()
    inputs = tuple(t.double() if isinstance(t, torch.Tensor) and t.is_floating_point() else t for t in inputs)

    with torch.no_grad():
        eager_output = model(*inputs)
    print(f'eager float64 output: {eager_output}')

    torch.manual_seed(42)
    model_fp32 = copy.deepcopy(model)
    model_fp32 = model_fp32.float()
    model_fp32.eval()

    inputs_fp32 = tuple(t.float() if isinstance(t, torch.Tensor) and t.is_floating_point() else t for t in inputs)
    compiled_model = torch.compile(model_fp32, mode="reduce-overhead")

    with torch.no_grad():
        compiled_output = compiled_model(*inputs_fp32)
    print(f'compile float32 output: {compiled_output}')

    abs_diff = (eager_output - compiled_output.double()).abs()
    rel_diff = abs_diff / (eager_output.abs() + 1e-8)
    is_inconsistent = (abs_diff > 0.01) | (rel_diff > 0.001)

    print(f'Abs diff max: {abs_diff.max().item():.6e}, mean: {abs_diff.mean().item():.6e}')
    print(f'Rel diff max: {rel_diff.max().item():.6e}, mean: {rel_diff.mean().item():.6e}')
    print(f'Inconsistent: {is_inconsistent.sum().item()} / {eager_output.numel()}')

if __name__ == '__main__':
    main()
```

### Error logs
output:
```
eager float64 output: tensor([[[ 8.1859e-02,  3.4557e-02, -6.3234e-02,  ..., -4.8920e-02,
           1.0489e-01, -1.1216e-03],
         [ 5.1179e-02,  4.7318e-02, -4.2696e-02,  ..., -8.6015e-02,
           1.3042e-01,  1.9244e-02],
         [ 1.0081e-02,  1.6891e-03,  9.1552e-03,  ..., -8.0940e-02,
           9.9519e-02,  1.7427e-02],
         ...,
         [ 3.1952e-02,  2.8212e-02, -3.9268e-02,  ..., -8.3000e-02,
           1.3541e-01,  4.1919e-02],
         [-1.1569e-03, -6.8458e-03, -3.7595e-02,  ..., -4.9887e-02,
           7.2429e-02, -3.6655e-03],
         [-8.3792e-03,  6.5382e-06, -1.1949e-02,  ..., -5.8032e-02,
           2.8549e-02,  1.5398e-02]],

        [[ 1.1317e-01,  2.5020e-03,  3.9420e-02,  ..., -5.2147e-02,
           8.2579e-02,  2.9163e-02],
         [ 6.0666e-02, -2.5938e-03,  4.3203e-02,  ..., -1.0380e-02,
           5.5590e-02,  7.6359e-03],
         [ 1.2379e-01, -2.4318e-02,  7.7615e-02,  ..., -1.8434e-02,
           5.5233e-02, -3.7587e-03],
         ...,
         [ 8.1636e-02, -3.5400e-02,  6.3066e-02,  ..., -1.3606e-01,
           3.0053e-02, -6.3494e-02],
         [ 7.8785e-02, -3.3025e-02,  1.2501e-02,  ..., -5.3827e-02,
           7.3924e-02, -1.2888e-02],
         [ 1.2321e-01, -3.0211e-02,  5.8782e-02,  ..., -6.6239e-02,
           1.0208e-01,  1.6668e-02]],

        [[-5.0298e-02, -4.6114e-02, -4.0889e-02,  ...,  5.4300e-02,
           4.3927e-02, -3.2966e-02],
         [-3.3095e-02,  2.4302e-02, -3.9467e-03,  ...,  5.3169e-02,
           3.4210e-02, -5.4969e-02],
         [-1.6826e-02, -1.0386e-02, -3.4127e-02,  ...,  3.1437e-02,
           3.6107e-02, -6.6378e-02],
         ...,
         [-3.8262e-02, -8.2813e-03, -1.1991e-03,  ...,  5.2024e-03,
           1.5615e-02, -3.3593e-02],
         [-3.2166e-02, -1.9082e-02, -3.7269e-03,  ..., -3.1149e-03,
           4.0433e-02, -2.9964e-02],
         [-1.8548e-02, -3.3213e-02, -2.3333e-02,  ...,  1.4079e-02,
           2.2524e-02, -9.2520e-03]],

        [[ 8.6348e-02,  1.4065e-02, -4.6246e-02,  ..., -6.3562e-03,
           3.0330e-02, -9.0822e-02],
         [ 1.0728e-01, -3.3761e-02, -7.7139e-02,  ..., -2.2990e-02,
           2.0806e-02, -7.5261e-02],
         [ 8.1223e-02,  1.0112e-02, -6.5384e-02,  ..., -2.9850e-02,
           4.4309e-02, -8.5228e-02],
         ...,
         [ 1.0010e-01, -7.8173e-04, -7.2807e-02,  ..., -2.6328e-02,
           3.9388e-02, -3.9802e-02],
         [ 8.7046e-02, -1.9918e-02, -8.4164e-02,  ..., -5.0626e-02,
           4.6767e-02, -5.1821e-02],
         [ 1.5549e-01,  3.1886e-02, -1.1294e-01,  ..., -1.6613e-02,
           4.5977e-02, -9.0721e-02]]], dtype=torch.float64)
compile float32 output: tensor([[[ 0.0532,  0.0100, -0.0736,  ..., -0.0428,  0.0916,  0.0217],
         [ 0.0368,  0.0319, -0.0682,  ..., -0.0631,  0.1189,  0.0369],
         [ 0.0031,  0.0029, -0.0234,  ..., -0.0617,  0.0994,  0.0476],
         ...,
         [ 0.0315,  0.0443, -0.0203,  ..., -0.0606,  0.1070,  0.0218],
         [-0.0181,  0.0089, -0.0191,  ..., -0.0362,  0.1221, -0.0088],
         [ 0.0270,  0.0158, -0.0232,  ..., -0.0776,  0.0369,  0.0034]],

        [[ 0.0917,  0.0245,  0.0298,  ..., -0.0640,  0.0675,  0.0353],
         [ 0.0955,  0.0076,  0.0509,  ..., -0.0580,  0.0469, -0.0084],
         [ 0.0829, -0.0163,  0.0620,  ..., -0.0267,  0.0646, -0.0216],
         ...,
         [ 0.0662, -0.0304,  0.0589,  ..., -0.1106,  0.0293, -0.0585],
         [ 0.1232, -0.0576,  0.0531,  ..., -0.0349,  0.0565, -0.0134],
         [ 0.1279, -0.0095,  0.0684,  ..., -0.0586,  0.0928,  0.0051]],

        [[-0.0496, -0.0487, -0.0342,  ...,  0.0340,  0.0306, -0.0391],
         [-0.0007,  0.0201, -0.0354,  ...,  0.0323,  0.0434, -0.0529],
         [-0.0142, -0.0348, -0.0061,  ...,  0.0609,  0.0424, -0.0451],
         ...,
         [ 0.0120, -0.0230, -0.0199,  ...,  0.0370,  0.0590, -0.0319],
         [-0.0622, -0.0114, -0.0055,  ...,  0.0379,  0.0614, -0.0269],
         [-0.0093, -0.0144,  0.0030,  ...,  0.0076,  0.0425, -0.0183]],

        [[ 0.0809,  0.0335, -0.0396,  ..., -0.0201,  0.0366, -0.0897],
         [ 0.1105, -0.0140, -0.0983,  ..., -0.0348,  0.0138, -0.0577],
         [ 0.0719, -0.0145, -0.0624,  ..., -0.0233,  0.0296, -0.0988],
         ...,
         [ 0.1222,  0.0057, -0.0606,  ..., -0.0402,  0.0155,  0.0006],
         [ 0.0870,  0.0140, -0.0490,  ..., -0.0444,  0.0261, -0.0311],
         [ 0.1233,  0.0107, -0.1301,  ..., -0.0089,  0.0452, -0.1020]]])
Abs diff max: 9.871764e-02, mean: 1.668783e-02
Rel diff max: 4.302332e+04, mean: 3.220054e+00
Inconsistent: 130846 / 131072
```
### Versions

PyTorch version: 2.10.0+cu126

OS: Ubuntu 24.04.1 LTS

GPU: NVIDIA GeForce RTX 4060 Laptop GPU

CUDA version: 12.6

Python version: 3.12.3

cc @pbelevich @mruberry @kurtamohler @chauhang @penguinwu @drisspg @liangel-02 @howardzhang-cv

**Comment by jansel** [comment]
I suspect this due to the `dropout_p=0.1` with different seeds.  If you disable dropout does this still repro?

**Comment by malfet** [comment]
Closing as won't-fix: this is expected behavior, not a `torch.compile` correctness bug.

The repro compares an **eager float64** run against a **compiled float32** run (`model.double()` vs `model_fp32.float()`), then flags any deviation above a hand-picked `abs > 0.01 | rel > 0.001`. That tolerance is tighter than float32 can deliver through a multi-head attention chain, so the difference is dtype, not compilation. See [Numerical accuracy](https://docs.pytorch.org/docs/stable/notes/numerical_accuracy.html): "PyTorch is not guaranteed to produce bitwise identical results for floating point computations that are mathematically identical."

There is also an RNG confound. `dropout_p=0.1` is passed directly to `F.scaled_dot_product_attention`, which `model.eval()` does **not** disable — it is a functional argument, not an `nn.Dropout` module. The two paths reach their forward at different positions in the RNG stream: the eager forward runs after model construction and `get_sample_inputs()` have drawn from the generator, while the compiled forward runs immediately after a bare `manual_seed(42)` and a `copy.deepcopy` (which draws nothing). The dropout masks therefore differ between the two runs.

`torch._inductor.config.fallback_random = True` aligns RNG *within* compiled regions; it does not make two differently-positioned draws agree. Identical results are not guaranteed across execution modes — see the [Reproducibility](https://docs.pytorch.org/docs/stable/notes/randomness.html) note.

To isolate `torch.compile`, run both paths in the same dtype with the same weights, and reset the seed immediately before each forward. #188425 was closed for the same reason.
